### DISCUSSION TITLE
Permission denied error

### DISCUSSION THREAD

**Discussion by Nu-hin** [firefly-iii/firefly-iii#9579] [Firefly III] [answered]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [x] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [x] ... there are no open or closed issues that are related to my problem
- [x] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

* **What's wrong:** Every now and then opening a page results in a 500 Server error page (error below).
* **Which versions are affected:** I observed it on 6.1.24 and 6.1.25.

The error is always with regards to the permission denied to access cache:

```
file_put_contents(/var/www/html/storage/framework/cache/data/aa/bb/aabbXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
Failed to open stream: Permission denied.
```

where `aa`, `bb`, and `aabbXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX` can be different.

Restarting the Docker container so far has helped fix the issue.

### Debug information

Debug information generated at 2024-12-25 09:06:48 for Firefly III version **v6.1.25**.

<table><thead><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Firefly III</td><td>6.1.25 / v2.1.0 / 25 (exp. 25)</td></tr><tr><td>PHP version</td><td>8.3.13 (64bits) / apache2handler / Linux x86_64</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Error reporting</td><td>Display: Off, reporting: ALL errors</td></tr><tr><td>Max upload</td><td>67108864 (64 MB)</td></tr><tr><td>Database drivers</td><td>mysql, <strong>*pgsql*</strong>, sqlite, </td></tr><tr><td>Docker build</td><td><span>#</span>1160, base <span>#</span>92</td></tr></tbody></table><table><thead><tr><th colspan="2">Firefly III information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Timezone</td><td>UTC + Europe/Berlin</td></tr><tr><td>App environment</td><td>production, debug: false</td></tr><tr><td>Layout</td><td>v1</td></tr><tr><td>Logging</td><td>debug, stack / (empty)</td></tr><tr><td>Cache driver</td><td>file</td></tr><tr><td>Default language and locale</td><td>en_US + equal</td></tr><tr><td>Trusted proxies</td><td>**</td></tr><tr><td>Login provider & user guard</td><td>eloquent / web</td></tr><tr><td>Login headers</td><td>N/A + N/A</td></tr><tr><td>Stateful domains</td><td></td></tr><tr><td>Last cron job</td><td>2024-12-24 23:00:01 (10 hours ago)</td></tr><tr><td>Mailer</td><td>log</td></tr></tbody></table><table><thead><tr><th colspan="2">User-specific information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>User</td><td><span>#</span>1 of 1</td></tr><tr><td>User flags</td><td><span title="Has reconciled">:ledger:</span> <span title="Has recurring transactions">:clock130:</span> <span title="Has object groups">:bookmark_tabs:</span> <span title="Has subscriptions">:email:</span></td></tr><tr><td>Session start</td><td>2024-12-01 00:00:00</td></tr><tr><td>Session end</td><td>2024-12-31 23:59:59</td></tr><tr><td>View range</td><td>1M</td></tr><tr><td>User language</td><td>en_US</td></tr><tr><td>User locale</td><td>en_US</td></tr><tr><td>Locale(s) supported</td><td>en_US.utf8: :white_check_mark:<br>en_US.UTF-8: :white_check_mark:<br></td></tr><tr><td>User agent</td><td>Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36</td></tr></tbody></table>

### Additional info

Here's how I run the container (Ansible code)

```ansible
- name: "Create Firefly container"
  docker_container:
    name: 'firefly'
    image: "fireflyiii/core:version-6.1.25"
    restart_policy: always
    env:
      APP_KEY: 'my-app-key-here'
      APP_URL: 'https://{{ firefly_host }}.{{ domain }}'
      DB_HOST: postgres
      DB_PORT: "5432"
      DB_CONNECTION: pgsql
      DB_DATABASE: firefly
      DB_USERNAME: firefly
      DB_PASSWORD: '{{ firefly_postgres_password }}'
      TRUSTED_PROXIES: '**'
      APP_LOG_LEVEL: debug
      MAIL_MAILER: log
    networks:
      - name: service-bridge
    networks_cli_compatible: no
    ports:
      - "{{ firefly_port }}:8080"
```

I started investigating, and found that the folders in `/var/www/html/storage/framework/cache/data` causing the trouble are owned by `root` user.

```
/var/www/html/storage/framework/cache/data# ls -la
total 936
...
drwxr-xr-x 3 www-data www-data 4096 Dec 25 08:40 55
drwxr-xr-x 6 www-data www-data 4096 Dec 25 08:40 56
drwxr-xr-x 3 root     root     4096 Dec 25 08:01 57
drwxr-xr-x 3 root     root     4096 Dec 25 08:01 58
drwxr-xr-x 7 www-data www-data 4096 Dec 25 08:40 59
drwxr-xr-x 3 www-data www-data 4096 Dec 25 08:40 5a
...
```

**Comment by JC5** [comment, marked as answer]
Hi, thanks for starting a discussion.

https://docs.firefly-iii.org/references/faq/docker/#i-get-failed-to-open-stream-permission-denied-on-log-files

**Comment by Nu-hin** [comment]
@JC5 Thank you for pointing that out, and I am very sorry that I haven't found that myself in the FAQ.

However, I must point out, that the official page [How to configure CRON jobs](https://docs.firefly-iii.org/how-to/firefly-iii/advanced/cron/#from-outside-the-container-cli) gives an explicit docker command, which doesn't seem to specify `--user data-www`:

```
# cron job for Firefly III using Docker itself
0 3 * * * docker exec $(docker container ls -a -f name=firefly --format="{{.ID}}") /usr/local/bin/php /var/www/html/artisan firefly-iii:cron
```

Moreover, I _used to_ have a cron job with `--user data-www`, but after reading [this](https://docs.firefly-iii.org/how-to/firefly-iii/advanced/cron/#from-outside-the-container-cli), I have _removed_ the `user` argument, because I assumed it was no longer needed. 

Can it be that the documentation on Cron Jobs should be amended too?

**Reply by JC5** [reply]
Good point! I'll pick it up :)

**Comment by Nu-hin** [comment]
Thanks!
